rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,481 @@
+    TlsEndpointTag tls_endpoint_tag = 4;
+  }
+}
+
+```
+
+As an example, the following can be used to define a label template matching a
+user-specific gRPC confidentiality tag and extracting the byte-array content of
+the tag as a parameter named user_id:
+
+```toml
+
+[[label_template.confidentiality_tags]]
+[label_template.confidentiality_tags.tag.grpc_tag_template]
+user_id_parameter_name = "user_id"
+
+```
+
+Note: I am unsure about the exact TOML representation for this.
+
+Question: In general TOML is not great for deeply nested object hierarchies, so
+we should perhaps consider a different serialisation format. Prost does not
+support text-serialised protos. JSON is an option, but not that great for human
+readable config. YAML can handle the complex object hierarchies, but the
+specification is very complex and parsing YAML is considered risky due to
+potential of injection vulnerabilities that is common when trying to support
+dynamic local object creation as defined by the specification. The application
+configuration is in TOML format, so having to deal with multiple serialisation
+formats when creating an application is not a great developer experience. Is
+there a suitable format that will work for everything?
+
+### Invocation Routing
+
+The planner node will be responsible for creating the required pseudo nodes as
+well. In the specific cases of HTTP- and gRPC pseudo nodes each pseudo node will
+send all invocations over a single dedicated invocation channel. The planner
+node will, by default, have the handle to receive these invocations and
+therefore be responsible for routing them to the right Wasm nodes based on the
+associated labels (specifically the label associated with the request receiver
+channel should match the label of the target Wasm node). Since it would be
+possible for multiple Wasm nodes to share the same label, the planner node would
+need an indication of which Wasm node should receive invocations with a specific
+label on the request receiver channel.
+
+If more advanced routing is required, the planner node could forward the
+invocation receiver channel to another node that can do the routing. The problem
+in this case is that the new node would have to be responsible for creating any
+per-label nodes as well if these nodes must be dynamically created, seeing that
+the planner node will no longer be able to see incoming invocations.
+
+Questions:
+
+- Would there be a requirement for multiple gRPC server pseudo nodes running on
+  different ports to each have a different configuration for Invocation routing?
+- Should we use special handling of channel types to determine the targets for
+  routing, or should we mark specific nodes as targeted invocation receivers?
+- Should we support overriding all routing so that the planner node still
+  receives all invocations, but always forward it to a single specified node
+  that is responsible for the routing?
```

Comment:
1) Not sure if there is a case where it's needed, but we shouldn't prohibit this.

2) I think choosing destinations solely based on labels may be non-deterministic (it's not obvious where the message will land). But if we will also use destination node names, then we don't need to search for nodes with right labels, since we will already just specify the destination.

3) Not sure what it means. What is _overriding all routing_ and where we want to override it?

Reply:
> 2. I think choosing destinations solely based on labels may be non-deterministic (it's not obvious where the message will land). But if we will also use destination node names, then we don't need to search for nodes with right labels, since we will already just specify the destination.

Who/what will specify the destination node names? How will the router read these? The only information currenlty availabel to a public router node is the label of the invocation.

> 3. Not sure what it means. What is _overriding all routing_ and where we want to override it?

The idea is that the router stops doing the router and gives the responsibility to another node.

Reply:
I don't think there should be special roles like planner or router; the system should allow such patterns to emerge, but should not try to fit everything into patterns, otherwise it feels we will need to reinvent Wasm anyways.